packages/local-storage-discovery/src/index.ts

```diff
@@ -0,0 +1,154 @@
+import { createFromJSON } from "@libp2p/peer-id-factory";
+import { multiaddr } from "@multiformats/multiaddr";
+import {
+  type Libp2pComponents,
+  type LocalStoragePeerInfo,
+  Tags
+} from "@waku/interfaces";
+import { Logger } from "@waku/utils";
+
+const log = new Logger("peer-exchange-discovery");
+
+type LocalStorageDiscoveryOptions = {
+  tagValue?: number;
+  tagTTL?: number;
+};
+
+export const DEFAULT_PEER_EXCHANGE_TAG_NAME = Tags.PEER_EXCHANGE;
+const DEFAULT_PEER_EXCHANGE_TAG_VALUE = 50;
+const DEFAULT_PEER_EXCHANGE_TAG_TTL = 100_000_000;
+
+export class LocalStorageDiscovery
+  extends TypedEventEmitter<PeerDiscoveryEvents>
+  implements PeerDiscovery, Startable
+{
+  private isStarted: boolean;
+
+  constructor(
+    private readonly components: Libp2pComponents,
+    private readonly options?: LocalStorageDiscoveryOptions
+  ) {
+    super();
+    this.isStarted = false;
+  }
+  get [Symbol.toStringTag](): string {
+    return "@waku/local-storage-discovery";
+  }
+
+  async start(): Promise<void> {
+    if (this.isStarted) {
+      return;
+    }
+
+    log.info("Starting Local Storage Discovery");
+
+    // only listen to healthy connected peers that have identified themselves
+    this.components.events.addEventListener("peer:update", this.handleNewPeers);
+
+    // get all peers from local storage
+    const localStoragePeers = this.getPeersFromLocalStorage();
+
+    //  update the tags of the peers in the peer store & dispatch them
+    for (const { id: idStr, address } of localStoragePeers) {
+      const peerId = await createFromJSON({
+        id: idStr
+      });
+
+      await this.components.peerStore.save(peerId, {
+        tags: {
+          [DEFAULT_PEER_EXCHANGE_TAG_NAME]: {
+            value: this.options?.tagValue ?? DEFAULT_PEER_EXCHANGE_TAG_VALUE,
+            ttl: this.options?.tagTTL ?? DEFAULT_PEER_EXCHANGE_TAG_TTL
+          }
+        }
+      });
+
+      this.dispatchEvent(
+        new CustomEvent<PeerInfo>("peer", {
+          detail: {
+            id: peerId,
+            multiaddrs: [multiaddr(address)]
+          }
+        })
+      );
+    }
+
+    log.info(`Discovered ${localStoragePeers.length} peers`);
+  }
+
+  stop(): void | Promise<void> {
+    if (!this.isStarted) return;
+    log.info("Stopping Local Storage Discovery");
+    this.components.events.removeEventListener(
+      "peer:update",
+      this.handleNewPeers
+    );
+    this.isStarted = false;
+  }
+
+  handleNewPeers = (event: CustomEvent<PeerUpdate>): void => {
+    const { peer } = event.detail;
+
+    // Extract the websocket ma
+    const websocketMultiaddr = peer.addresses.find((addr) =>
+      addr.toString().includes("ws" || "wss")
+    );
+    if (!websocketMultiaddr) return;
+
+    // Retrieve peers from local storage
+    const storedPeersData = localStorage.getItem("waku:peers");
+    const localStoragePeers = (
+      storedPeersData ? JSON.parse(storedPeersData) : []
+    ) as LocalStoragePeerInfo[];
```

Comment:
This code is not needed here as peers should be read from `localStorage` only in `start` and then maintained by this event.

Reply:
can you please elaborate?

this is needed to store new discovered peers (through peer-exchange) in local storage that can be used the next time app (js-waku node) is started

Reply:
Idea is - we read peers from store only at the start of the service. It goes to a map or any other DS and maintained when needed. At the end or in the process backup is done.

Here you do get it every time which might work but I think conceptually wrong.

Reply:
How would you define `At the end or in the process backup is done.`?
I would think updating each time directly in the storage is the most reliable way
Happy to discuss & address as a followup as well

Reply:
It can be some kind of event either from `libp2p` or html `beforeunload` event
for offloading it is not critical but for reading I don't think it is good to do on each event

Reply:
hmm curious how you see it -- would you be willing to share a code snipper/open a PR? (can be done as a followup as well)

Reply:
it seems `stop` method can be a good place - but it is more of a nice-to-have thing since you fixed reading on each iteration - which I think is more critical than writing on each iteration

Reply:
sure, I think about something like:
```ts
export class LocalStorageDiscovery
  extends TypedEventEmitter<PeerDiscoveryEvents>
  implements PeerDiscovery, Startable
{
  private isStarted: boolean;
  private peers: LocalStoragePeerInfo[] = [];

  constructor(
    private readonly components: Libp2pComponents,
    private readonly options?: LocalStorageDiscoveryOptions
  ) {
    super();
    this.isStarted = false;
  }

  get [Symbol.toStringTag](): string {
    return "@waku/local-discovery";
  }

  async start(): Promise<void> {
    if (this.isStarted) {
      return;
    }

    log.info("Starting Local Storage Discovery");

    this.components.events.addEventListener(
      "peer:identify",
      this.handleNewPeers
    );

    this.peers = this.getPeersFromLocalStorage();

    for (const { id: idStr, address } of localStoragePeers) {
      const peerId = await createFromJSON({
        id: idStr
      });

      if (await this.components.peerStore.has(peerId)) return;

      await this.components.peerStore.save(peerId, {
        multiaddrs: [multiaddr(address)],
        tags: {
          [DEFAULT_LOCAL_TAG_NAME]: {
            value: this.options?.tagValue ?? DEFAULT_LOCAL_TAG_VALUE,
            ttl: this.options?.tagTTL ?? DEFAULT_LOCAL_TAG_TTL
          }
        }
      });

      this.dispatchEvent(
        new CustomEvent<PeerInfo>("peer", {
          detail: {
            id: peerId,
            multiaddrs: [multiaddr(address)]
          }
        })
      );
    }

    log.info(`Discovered ${localStoragePeers.length} peers`);
  }

  stop(): void | Promise<void> {
    if (!this.isStarted) return;
    log.info("Stopping Local Storage Discovery");
    this.components.events.removeEventListener(
      "peer:identify",
      this.handleNewPeers
    );
    this.isStarted = false;
    // might be the other way to handle backup gracefully
    this.setPeersInLocalStorage(this.peers);
  }

  handleNewPeers = (event: CustomEvent<IdentifyResult>): void => {
    const peer = event.detail;

    const websocketMultiaddr = getWsMultiaddrFromMultiaddrs(peer.listenAddrs);

    const existingPeerIndex = peers.findIndex(
      (_peer) => _peer.id === peer.peerId.toString()
    );

    if (existingPeerIndex >= 0) {
      peers[existingPeerIndex].address =
        websocketMultiaddr.toString();
    } else {
      peers.push({
        id: peer.peerId.toString(),
        address: websocketMultiaddr.toString()
      });
    }

    // if not possible to backup in `stop` method
    this.setPeersInLocalStorage(peers);
  };

  private getPeersFromLocalStorage(): LocalStoragePeerInfo[] {
    try {
      const storedPeersData = localStorage.getItem("waku:peers");
      if (!storedPeersData) return [];
      const peers = JSON.parse(storedPeersData);
      return peers.filter(isValidStoredPeer);
    } catch (error) {
      log.error("Error parsing peers from local storage:", error);
      return [];
    }
  }

  private setPeersInLocalStorage(peers: LocalStoragePeerInfo[]): void {
    localStorage.setItem("waku:peers", JSON.stringify(peers));
  }
}
```